## Section 1: Basic Decoding Algorithms

In [1]:
!uv pip install transformers datasets evaluate markdownify

Using Python 3.11.12 environment at: /usr
Resolved 39 packages in 510ms
⠙ Preparing packages... (0/7)
⠙ Preparing packages... (0/7)
dill       ------------------------------     0 B/113.53 KiB
⠙ Preparing packages... (0/7)
markdownify ------------------------------     0 B/13.58 KiB
dill       ------------------------------     0 B/113.53 KiB
⠙ Preparing packages... (0/7)
markdownify ------------------------------     0 B/13.58 KiB
dill       ------------------------------ 14.93 KiB/113.53 KiB
⠙ Preparing packages... (0/7)
markdownify ------------------------------ 13.58 KiB/13.58 KiB
dill       ------------------------------ 14.93 KiB/113.53 KiB
⠙ Preparing packages... (0/7)
dill       ------------------------------ 14.93 KiB/113.53 KiB
⠙ Preparing packages... (0/7)
dill       ------------------------------ 14.93 KiB/113.53 KiB
multiprocess ------------------------------     0 B/140.16 KiB
⠙ Preparing packages... (0/7)
dill       ------------------------------ 14.93 KiB/113.53 KiB
mul

In [2]:
"""set device and random seeds"""

######################################################
#  The following helper functions are given to you.
######################################################

from tqdm.notebook import tqdm
import torch
import torch.nn.functional as F

device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(f'device: {device}')

def set_seed(seed=19260817):
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

set_seed()

device: cuda


In [3]:
"""load datasets"""

######################################################
#  The following helper code is given to you.
######################################################

from datasets import load_dataset

dataset = load_dataset('Ximing/ROCStories')
train_data, dev_data, test_data = dataset['train'], dataset['validation'], dataset['test']

print(train_data[0])

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/131 [00:00<?, ?B/s]

train.jsonl:   0%|          | 0.00/6.53M [00:00<?, ?B/s]

validation.jsonl:   0%|          | 0.00/717k [00:00<?, ?B/s]

test.jsonl:   0%|          | 0.00/713k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/16207 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/1817 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/1819 [00:00<?, ? examples/s]

{'story_id': '080198fc-d0e7-42b3-8e63-b2144e59d816', 'prompt': 'On my way to work I stopped to get some coffee.', 'continuation': 'I went through the drive through and placed my order. I paid the cashier and patiently waited for my drink. When she handed me the drink, the lid came off and spilled on me. The coffee hurt and I had to go home and change clothes.', 'constraint_words': ['drive', 'order', 'drink', 'lid', 'coffee', 'hurt', 'home', 'change', 'clothes']}


In [4]:
"""prepare evaluation"""

######################################################
#  The following helper code is given to you.
######################################################

from evaluate import load
from transformers import RobertaForSequenceClassification, RobertaTokenizer

perplexity_scorer = load("perplexity", module_type="metric")
cola_model_name = "textattack/roberta-base-CoLA"
cola_tokenizer = RobertaTokenizer.from_pretrained(cola_model_name)
cola_model = RobertaForSequenceClassification.from_pretrained(cola_model_name).to(device)

def batchify(data, batch_size):
    assert batch_size > 0

    batch = []
    for item in data:
        # Yield next batch
        if len(batch) == batch_size:
            yield batch
            batch = []

        batch.append(item)

    # Yield last un-filled batch
    if len(batch) != 0:
        yield batch

tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/564 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/501M [00:00<?, ?B/s]

Some weights of the model checkpoint at textattack/roberta-base-CoLA were not used when initializing RobertaForSequenceClassification: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [5]:
"""set up evaluation metric"""

######################################################
#  The following helper code is given to you.
######################################################

def compute_perplexity(texts, model='gpt2', batch_size=8):
    score = perplexity_scorer.compute(predictions=texts, add_start_token=True, batch_size=batch_size, model_id=model)
    return score['mean_perplexity']


def compute_fluency(texts, batch_size=8):
  scores = []
  for b_texts in batchify(texts, batch_size):
    inputs = cola_tokenizer(texts, padding=True, truncation=True, return_tensors="pt").to(device)
    with torch.no_grad():
      logits = cola_model(**inputs).logits
      probs = logits.softmax(dim=-1)
      scores.extend(probs[:, 1].tolist())
  return sum(scores) / len(scores)


def compute_diversity(texts):
    unigrams, bigrams, trigrams = [], [], []
    total_words = 0
    for gen in texts:
        o = gen.split(' ')
        total_words += len(o)
        for i in range(len(o)):
            unigrams.append(o[i])
        for i in range(len(o) - 1):
            bigrams.append(o[i] + '_' + o[i + 1])
        for i in range(len(o) - 2):
            trigrams.append(o[i] + '_' + o[i + 1] + '_' + o[i + 2])
    return len(set(unigrams)) / len(unigrams), len(set(bigrams)) / len(bigrams), len(set(trigrams)) / len(trigrams)


def evaluate(generations, experiment):
  generations = [_ for _ in generations if _ != '']
  perplexity = compute_perplexity(generations)
  fluency = compute_fluency(generations)
  diversity = compute_diversity(generations)
  print(experiment)
  print(f'perplexity = {perplexity:.2f}')
  print(f'fluency = {fluency:.2f}')
  print(f'diversity = {diversity[0]:.2f}, {diversity[1]:.2f}, {diversity[2]:.2f}')
  print()

debug_sents = ["This restaurant is awesome", "My dog is cute and I love it.", "Today is sunny."]
evaluate(debug_sents, 'debugging run')

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

  0%|          | 0/1 [00:00<?, ?it/s]

debugging run
perplexity = 178.64
fluency = 0.98
diversity = 0.87, 1.00, 1.00



In [6]:
"""load model and tokenizer"""

######################################################
#  The following helper code is given to you.
######################################################

from transformers import GPT2LMHeadModel, GPT2Tokenizer

model_name = 'gpt2'
tokenizer = GPT2Tokenizer.from_pretrained(model_name, pad_token="<|endoftext|>")
model = GPT2LMHeadModel.from_pretrained(model_name).to(device)
model.eval()

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=2304, nx=768)
          (c_proj): Conv1D(nf=768, nx=768)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=3072, nx=768)
          (c_proj): Conv1D(nf=768, nx=3072)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

In this section, you will implement a few basic decoding algorithms:
1. Greedy decoding
2. Vanilla sampling
3. Temperature sampling
4. Top-k sampling
5. Top-p sampling

We have provided a wrapper function `decode()` that takes care of batching, controlling max length, and handling the EOS token.
You will be asked to implement the core function of each method: *given the pre-softmax logits of the next token, decide what the next token is.*

**The wrapper calls the core function of each decoding algorithm, which you will implement in the subsections below.**

In [7]:
"""decode main wrapper function"""

######################################################
#  The following helper code is given to you.
######################################################

def decode(prompts, max_len, method, **kwargs):
  encodings_dict = tokenizer(prompts, return_tensors="pt", padding=True)
  input_ids = encodings_dict['input_ids'].to(device)
  attention_mask = encodings_dict['attention_mask'].to(device)

  model_kwargs = {'attention_mask': attention_mask, "use_cache": False}
  batch_size, input_seq_len = input_ids.shape

  unfinished_sequences = torch.ones(batch_size, dtype=torch.long, device=device)

  for step in range(max_len):
    model_inputs = model.prepare_inputs_for_generation(input_ids, **model_kwargs)
    cache_position = torch.arange(0, input_ids.shape[1], dtype=torch.long, device=input_ids.device)
    model_kwargs["cache_position"] = cache_position
    with torch.no_grad():
      outputs = model(**model_inputs, return_dict=True, output_attentions=False, output_hidden_states=False)

    if step == 0:
      last_non_masked_idx = torch.sum(attention_mask, dim=1) - 1
      next_token_logits = outputs.logits[range(batch_size), last_non_masked_idx, :]
    else:
      next_token_logits = outputs.logits[:, -1, :]

    log_prob = F.log_softmax(next_token_logits, dim=-1)

    if method == 'greedy':
      next_tokens = greedy(next_token_logits)
    elif method == 'sample':
      next_tokens = sample(next_token_logits)
    elif method == 'temperature':
      next_tokens = temperature(next_token_logits, t=kwargs.get('t', 0.8))
    elif method == 'topk':
      next_tokens = topk(next_token_logits, k=kwargs.get('k', 20))
    elif method == 'topp':
      next_tokens = topp(next_token_logits, p=kwargs.get('p', 0.7))

    # finished sentences should have their next token be a padding token
    next_tokens = next_tokens * unfinished_sequences + tokenizer.pad_token_id * (1 - unfinished_sequences)

    input_ids = torch.cat([input_ids, next_tokens[:, None]], dim=-1)
    model_kwargs = model._update_model_kwargs_for_generation(outputs, model_kwargs, is_encoder_decoder=model.config.is_encoder_decoder)

    # if eos_token was found in one sentence, set sentence to finished
    unfinished_sequences = unfinished_sequences.mul((next_tokens != tokenizer.eos_token_id).long())

    if unfinished_sequences.max() == 0:
      break

  response_ids = input_ids[:, input_seq_len:]
  response_text = [tokenizer.decode(output, skip_special_tokens=True, clean_up_tokenization_spaces=True) for output in response_ids]

  return response_text

In [8]:
"""debug helper code"""

######################################################
#  The following helper code is given to you.
######################################################

# For debugging, we duplicate a single prompt 10 times so that we obtain 10 generations for the same prompt
dev_prompts = [dev_data[0]['prompt']] * 10

def print_generations(prompts, generations):
  for prompt, generation in zip(prompts, generations):
    print(f'{[prompt]} ==> {[generation]}')

### 1.1 Greedy Decoding

In [9]:
def greedy(next_token_logits):
  '''
  inputs:
  - next_token_logits: Tensor(size = (B, V), dtype = float)
  outputs:
  - next_tokens: Tensor(size = (B), dtype = long)
  '''

  # TODO: compute `next_tokens` from `next_token_logits`.
  # Hint: use torch.argmax()
  next_tokens = torch.argmax(next_token_logits, dim=1)

  return next_tokens


generations = decode(dev_prompts, max_len=20, method='greedy')
print_generations(dev_prompts, generations)

['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I was like, \'I\'m going to go to work tomorrow,\'" he said.']
['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I was like, \'I\'m going to go to work tomorrow,\'" he said.']
['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I was like, \'I\'m going to go to work tomorrow,\'" he said.']
['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I was like, \'I\'m going to go to work tomorrow,\'" he said.']
['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I was like, \'I\'m going to go to work tomorrow,\'" he said.']
['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I was like, \'I\'m going to go to work tomorrow,\'" he said.']
['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I was like, \'I\'m going to go to work tomorrow,\'" he said.']
['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I was like, \'I\'m goin

### 1.2 Vanilla Sampling and Temperature Sampling

In [10]:
def sample(next_token_logits):
  '''
  inputs:
  - next_token_logits: Tensor(size = (B, V), dtype = float)
  outputs:
  - next_tokens: Tensor(size = (B), dtype = long)
  '''

  # TODO: compute the probabilities `probs` from the logits.
  # Hint: `probs` should have size (B, V)
  probs = torch.softmax(next_token_logits, dim=1)

  # TODO: compute `next_tokens` from `probs`.
  # Hint: use torch.multinomial()
  next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)

  return next_tokens


set_seed()
generations = decode(dev_prompts, max_len=20, method='sample')
print_generations(dev_prompts, generations)

['Ryan was called by his friend to skip work one day.'] ==> ['\n\nKiddiak said he stepped on it, hit the brakes hard and decided to cross first']
['Ryan was called by his friend to skip work one day.'] ==> [" Their interaction didn't work at all by the time he would get back to Louisville. Some details that"]
['Ryan was called by his friend to skip work one day.'] ==> [' According to investigators at the time, Gianfranco was yelling at them for not doing enough work']
['Ryan was called by his friend to skip work one day.'] ==> [' "No, my boyfriend doesn\'t do work," Cat started, desperate to dial a phone in distress']
['Ryan was called by his friend to skip work one day.'] ==> [' Later, he did not show up for school again until late the next day. After catching a train']
['Ryan was called by his friend to skip work one day.'] ==> [' The fliers denounced Gov. Pat Brown for having insufficient police presence in the building, and he engineered']
['Ryan was called by his friend to skip wo

In [11]:
def temperature(next_token_logits, t):
  '''
  inputs:
  - next_token_logits: Tensor(size = (B, V), dtype = float)
  - t: float
  outputs:
  - next_tokens: Tensor(size = (B), dtype = long)
  '''

  # TODO: compute the probabilities `probs` from the logits, with temperature applied.
  probs = torch.softmax(next_token_logits / t, dim=1)

  # TODO: compute `next_tokens` from `probs`.
  next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)

  return next_tokens


set_seed()
generations = decode(dev_prompts, max_len=20, method='temperature', t=0.8)
print_generations(dev_prompts, generations)

['Ryan was called by his friend to skip work one day.'] ==> ['\n\nKiddiak said he stepped on the blood spur out of pity from his cousin Jimmie']
['Ryan was called by his friend to skip work one day.'] ==> [" Their interaction didn't go well.\n\nHerad's mother was in a panic. Unable to"]
['Ryan was called by his friend to skip work one day.'] ==> [" He said he hadn't asked if he'd be coming back to work for a week.\n\n"]
['Ryan was called by his friend to skip work one day.'] ==> [' "I told my boss I was leaving work early to take care of my sons\' homework. So']
['Ryan was called by his friend to skip work one day.'] ==> [' He also said he was going to play out of town because the game was cancelled. Johnson also said']
['Ryan was called by his friend to skip work one day.'] ==> [' The next morning, after talking to other parents who also work in the area, he was told to']
['Ryan was called by his friend to skip work one day.'] ==> [' It was before he was going to college.\n\n"Oh, I\'l

### 1.3 Top-k Sampling

In [12]:
def topk(next_token_logits, k):
  '''
  inputs:
  - next_token_logits: Tensor(size = (B, V), dtype = float)
  - k: int
  outputs:
  - next_tokens: Tensor(size = (B), dtype = long)
  '''

  # TODO: Keep only top-k tokens with highest probabilities.
  # Hint: use torch.topk()
  topk_logits, topk_indices = torch.topk(next_token_logits, k, dim=1)

  # Create a mask to zero out all logits not in top-k
  indices_to_remove = next_token_logits < topk_logits[:, -1].unsqueeze(1)

  # Mask the logits
  next_token_logits[indices_to_remove] = float('-inf')

  # TODO: Sample from the masked logits
  probs = torch.softmax(next_token_logits, dim=1)
  next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)

  return next_tokens


set_seed()
generations = decode(dev_prompts, max_len=20, method='topk', k=20)
print_generations(dev_prompts, generations)

['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I wasn\'t sure what to expect, so he asked me why I didn\'t work']
['Ryan was called by his friend to skip work one day.'] ==> ['\n\n"I\'m tired of talking about this because it\'s a big distraction. I\'m tired']
['Ryan was called by his friend to skip work one day.'] ==> [' When he arrived at work, the other workers were on the line and asked if he would be willing']
['Ryan was called by his friend to skip work one day.'] ==> [' "Why don\'t you try to do work? You don\'t look like a person," his friend']
['Ryan was called by his friend to skip work one day.'] ==> [' He said he did not have enough money to pay his bills. He told his girlfriend, a teacher']
['Ryan was called by his friend to skip work one day.'] ==> [' The next morning, after work, he would get back to his job, and the next day,']
['Ryan was called by his friend to skip work one day.'] ==> [" It was the first time they had ever met, and they couldn't qui

### 1.4 Top-p Sampling

In [13]:
def topp(next_token_logits, p):
  '''
  inputs:
  - next_token_logits: Tensor(size = (B, V), dtype = float)
  - p: float
  outputs:
  - next_tokens: Tensor(size = (B), dtype = long)
  '''

  # TODO: Sort the logits in descending order, and compute
  # the cumulative probabilities `cum_probs` on the sorted logits
  sorted_logits, sorted_indices = torch.sort(next_token_logits, dim=1, descending=True)
  sorted_probs = torch.softmax(sorted_logits, dim=1)
  cum_probs = torch.cumsum(sorted_probs, dim=1)

  # Create a mask to zero out all logits not in top-p
  sorted_indices_to_remove = cum_probs > p
  sorted_indices_to_remove[:, 1:] = sorted_indices_to_remove[:, :-1].clone()
  sorted_indices_to_remove[:, 0] = 0
  # Restore mask to original indices
  indices_to_remove = sorted_indices_to_remove.scatter(dim=1, index=sorted_indices, src=sorted_indices_to_remove)

  # Mask the logits
  next_token_logits[indices_to_remove] = float('-inf')

  # TODO: Sample from the masked logits
  probs = torch.softmax(next_token_logits, dim=1)
  next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)

  return next_tokens


set_seed()
generations = decode(dev_prompts, max_len=20, method='topp', p=0.7)
print_generations(dev_prompts, generations)

['Ryan was called by his friend to skip work one day.'] ==> ["\n\nKiddie said he also didn't know if he was working out or not, and"]
['Ryan was called by his friend to skip work one day.'] ==> [' On the day of the attack, he took the opportunity to share a beautiful moment with her, a']
['Ryan was called by his friend to skip work one day.'] ==> [' When he arrived at work, the new tenant was coming in with a pile of trash.\n\n']
['Ryan was called by his friend to skip work one day.'] ==> [' "No, my boyfriend doesn\'t do that," he said, before jogging out to his room']
['Ryan was called by his friend to skip work one day.'] ==> [' He also sat in a wheelchair at his hospital bed. In the room was a young woman who looked']
['Ryan was called by his friend to skip work one day.'] ==> [" The rookie's parents said he was talking to his father and asking him to help him.\n\n"]
['Ryan was called by his friend to skip work one day.'] ==> [" Then he goes back to work. He can't afford to do it. S

### 1.5: Evaluation

Run the following cell to obtain the evaluation results, which you should include in your writeup.
Also don't forget to answer the questions.

In [14]:
prompts = [item['prompt'] for item in test_data][:10]
GENERATIONS_PER_PROMPT = 10
MAX_LEN = 100

for experiment in ['greedy', 'sample', 'temperature', 'topk', 'topp']:
  generations = []
  for prompt in tqdm(prompts):
    generations += decode([prompt] * GENERATIONS_PER_PROMPT, max_len=MAX_LEN, method=experiment)
  evaluate(generations, experiment)

  0%|          | 0/10 [00:00<?, ?it/s]

  0%|          | 0/13 [00:00<?, ?it/s]

greedy
perplexity = 2.08
fluency = 0.78
diversity = 0.01, 0.02, 0.03



  0%|          | 0/10 [00:00<?, ?it/s]

  0%|          | 0/13 [00:00<?, ?it/s]

sample
perplexity = 70.58
fluency = 0.34
diversity = 0.43, 0.90, 0.99



  0%|          | 0/10 [00:00<?, ?it/s]

  0%|          | 0/13 [00:00<?, ?it/s]

temperature
perplexity = 16.61
fluency = 0.68
diversity = 0.33, 0.79, 0.95



  0%|          | 0/10 [00:00<?, ?it/s]

  0%|          | 0/13 [00:00<?, ?it/s]

topk
perplexity = 12.24
fluency = 0.74
diversity = 0.26, 0.75, 0.96



  0%|          | 0/10 [00:00<?, ?it/s]

  0%|          | 0/13 [00:00<?, ?it/s]

topp
perplexity = 12.40
fluency = 0.72
diversity = 0.29, 0.76, 0.96



### Discussion
- Q1.1: In greedy decoding, what do you observe when generating 10 times from the test prompt?
    - All 10 generations are exactly identical. This is because the model consistently selects the token with the highest probability at each step without introducing any randomness.

- Q1.2: In vanilla sampling, what do you observe when generating 10 times from the test prompt?
    - The 10 generations are highly diverse, with each output being different in content, structure and even coherence. It seems to me that most of the continuations bear little meaningful connection to the given prompt.

- Q1.3: In temperature sampling, play around with the value of temperature $t$. Which value of $t$ makes it equivalent to greedy decoding? Which value of $t$ makes it equivalent to vanilla sampling?
    - When $t\to 0$, the probability distribution becomes really sharp and concentrates nearly all the probability mass on the highest logit. In this case, the sampling almost always selects the top token, which behaves like **greedy encoding**.
    - When $t=1$, the raw probabilities are not scaled, which makes temperature sampling behave identically to **vanilla sampling.**

- Q1.4: In top-$p$ sampling, play around with the value of $p$. Which value of $p$ makes it equivalent to greedy decoding? Which value of $p$ makes it equivalent to vanilla sampling?
    - When $p\to 0$, only the single probable token is selected (we can observe from the code that ``sorted_indices_to_remove[:, 0] = 0`` ensures that the token with highest probability is never removed), which makes top-p sampling equivalent to **greedy decoding**.
    - When $p=1$, the full probability mass is retained, thus the sampling behaves equivalently to **vanilla sampling**.

- Q1.5: In top-k sampling, play around with the value of k. Which value of k makes it equivalent to greedy decoding? Which value of $k$ makes it equivalent to vanilla sampling?
    - When $k=1$, the model always selects the token with the highest probability. There is no randomness left and it behaves exactly like **greedy decoding**.
    - When $k=|V|=50257$, the model retains all tokens for sampling. This is equivalent to **vanilla sampling**.

- Q1.6: Report the evaluation metrics (perplexity, fluency, diversity) of all 5 decoding methods. Which methods have the best and worst perplexity? Fluency? Diversity?


## Section 2: Applying Large Language Models to Few Shot Math Reasoning

In [ ]:
!uv pip install -q vllm # restart runtime session after install requirements

In [ ]:
import os
os.environ["HF_HOME"] = "/content/.cache/huggingface" # set the cache directory to personal disk to avoid frequent downloads

In [ ]:
from vllm import LLM, SamplingParams
model_id = "Qwen/Qwen2.5-7B-Instruct-GPTQ-Int4"
llm = LLM(model=model_id, enforce_eager=True)

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

from openai import OpenAI
from transformers import AutoTokenizer

class VLLMClient:
    def __init__(self, model_id, **kwargs):
      self.model_id = model_id

    def __call__(self, prompt: str, **kwargs):
      response = llm.generate(
            prompts=prompt,
            sampling_params=SamplingParams(
                temperature=kwargs.get("temperature", 0.2),
                max_tokens=kwargs.get("max_tokens", 256),
            )
      )
      return response[0].outputs[0].text
model = VLLMClient(model_id)
model("San Francisco is a", max_tokens=42)

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

GSM_EXAMPLARS = [
    {
        "question": "There are 15 trees in the grove. Grove workers will plant trees in the grove today. After they are done, there will be 21 trees. How many trees did the grove workers plant today?",
        "cot_answer": "There are 15 trees originally. Then there were 21 trees after some more were planted. So there must have been 21 - 15 = 6. So the answer is 6.",
        "short_answer": "6"
    },
    {
        "question": "If there are 3 cars in the parking lot and 2 more cars arrive, how many cars are in the parking lot?",
        "cot_answer": "There are originally 3 cars. 2 more cars arrive. 3 + 2 = 5. So the answer is 5.",
        "short_answer": "5"
    },
    {
        "question": "Leah had 32 chocolates and her sister had 42. If they ate 35, how many pieces do they have left in total?",
        "cot_answer": "Originally, Leah had 32 chocolates. Her sister had 42. So in total they had 32 + 42 = 74. After eating 35, they had 74 - 35 = 39. So the answer is 39.",
        "short_answer": "39"
    },
    {
        "question": "Jason had 20 lollipops. He gave Denny some lollipops. Now Jason has 12 lollipops. How many lollipops did Jason give to Denny?",
        "cot_answer": "Jason started with 20 lollipops. Then he had 12 after giving some to Denny. So he gave Denny 20 - 12 = 8. So the answer is 8.",
        "short_answer": "8"
    },
    {
        "question": "Shawn has five toys. For Christmas, he got two toys each from his mom and dad. How many toys does he have now?",
        "cot_answer": "Shawn started with 5 toys. If he got 2 toys each from his mom and dad, then that is 4 more toys. 5 + 4 = 9. So the answer is 9.",
        "short_answer": "9"
    },
    {
        "question": "There were nine computers in the server room. Five more computers were installed each day, from monday to thursday. How many computers are now in the server room?",
        "cot_answer": "There were originally 9 computers. For each of 4 days, 5 more computers were added. So 5 * 4 = 20 computers were added. 9 + 20 is 29. So the answer is 29.",
        "short_answer": "29"
    },
    {
        "question": "Michael had 58 golf balls. On tuesday, he lost 23 golf balls. On wednesday, he lost 2 more. How many golf balls did he have at the end of wednesday?",
        "cot_answer": "Michael started with 58 golf balls. After losing 23 on tuesday, he had 58 - 23 = 35. After losing 2 more, he had 35 - 2 = 33 golf balls. So the answer is 33.",
        "short_answer": "33"
    },
    {
        "question": "Olivia has $23. She bought five bagels for $3 each. How much money does she have left?",
        "cot_answer": "Olivia had 23 dollars. 5 bagels for 3 dollars each will be 5 x 3 = 15 dollars. So she has 23 - 15 dollars left. 23 - 15 is 8. So the answer is 8.",
        "short_answer": "8"
    }
]

In [ ]:
!mkdir -p data
!wget -q -O data/gsm8k.jsonl https://raw.githubusercontent.com/ranpox/comp3361-spring2025/refs/heads/main/assignments/A3/data/gsm8k.jsonl
!wget -q -O data/simpleqa.jsonl https://raw.githubusercontent.com/ranpox/comp3361-spring2025/refs/heads/main/assignments/A3/data/simpleqa.jsonl
!wget -q -O data/math.jsonl https://raw.githubusercontent.com/ranpox/comp3361-spring2025/refs/heads/main/assignments/A3/data/math.jsonl
!wget -q -O data/gaia.jsonl https://raw.githubusercontent.com/ranpox/comp3361-spring2025/refs/heads/main/assignments/A3/data/gaia.jsonl

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

import json

def load_eval_data(task):
    with open(f"data/{task}.jsonl", "r") as f:
        return [json.loads(line) for line in f]

tasks = ["gsm8k", "simpleqa", "math", "gaia"]

for task in tasks:
    print(f"Example of {task} dataset:")
    print(json.dumps(load_eval_data(task)[0], indent=4))

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

import os
import datetime
import threading
from pathlib import Path
from tqdm.notebook import tqdm
from concurrent.futures import ThreadPoolExecutor, as_completed

APPEND_ANSWER_LOCK = threading.Lock()

def append_answer(entry: dict, jsonl_file: str) -> None:
    jsonl_file = Path(jsonl_file)
    jsonl_file.parent.mkdir(parents=True, exist_ok=True)
    with APPEND_ANSWER_LOCK, open(jsonl_file, "a", encoding="utf-8") as fp:
        fp.write(json.dumps(entry) + "\n")
    assert os.path.exists(jsonl_file), "File not found!"

def answer_single_question(example, agent, answers_file, action_type):
    augmented_question = example["question"]
    if example["source"] == "SimpleQA":
        augmented_question += " Answer with only the final number."
    if example["source"] == "MATH":
        augmented_question += " Write code, not latex."

    start_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    answer = str(agent.run(augmented_question))

    end_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    annotated_example = {
        "model_id": model.model_id,
        "agent_action_type": action_type,
        "question": augmented_question,
        "original_question": example["question"],
        "answer": answer,
        "true_answer": example["true_answer"],
        "source": example["source"],
        "start_time": start_time,
        "end_time": end_time,
    }
    append_answer(annotated_example, answers_file)

def answer_questions(
    task,
    agent,
    action_type: str = "vanilla",
    answers_file: str = None,
    parallel_workers: int = 4,
):
    eval_data = load_eval_data(task)

    print(f"Starting processing and writing output to '{answers_file}'")

    answered_questions = []
    if os.path.exists(answers_file):
        with open(answers_file, "r") as f:
            for line in f:
                answered_questions.append(json.loads(line)["original_question"])

    examples_todo = [example for example in eval_data if example["question"] not in answered_questions]


    for i, example in enumerate(tqdm(examples_todo)):
        answer_single_question(example, agent, answers_file, action_type)

In [ ]:
!wget -q -O eval_utils.py https://raw.githubusercontent.com/ranpox/comp3361-spring2025/refs/heads/main/assignments/A3/evaluate.py

In [ ]:
from eval_utils import score_answers
class FewShotReasoner:
    def __init__(self, model, n_shots):
        self.model = model
        self.n_shots = n_shots

    def run(self, task):
        # TODO

    def build_input(self, task):
        # TODO


def run_gsm8k_fewshot_reasoner(task="gsm8k", model_id=model_id, action_type="vanilla"):
    reasoner_answers_file = f"output/{model_id.replace('/', '__')}__{action_type}__{task}.jsonl"
    chat_agent = FewShotReasoner(model, 8)
    answer_questions(task, chat_agent, action_type, reasoner_answers_file)
    df = score_answers([reasoner_answers_file])
    print(df)

run_gsm8k_fewshot_reasoner()

In [ ]:
import re

class FewShotCoTReasoner:
    def __init__(self, model, n_shots):
        self.model = model
        self.n_shots = n_shots

    def run(self, task):
        # TODO

    def build_input(self, task):
        # TODO


def run_gsm8k_fewshot_cot_reasoner(task="gsm8k", model_id=model_id, action_type="vanilla"):
    reasoner_answers_file = f"output/{model_id.replace('/', '__')}__{action_type}__{task}__cot.jsonl"
    chat_agent = FewShotCoTReasoner(model, 8)
    answer_questions(task, chat_agent, action_type, reasoner_answers_file)
    df = score_answers([reasoner_answers_file])
    print(df)

run_gsm8k_fewshot_cot_reasoner()

## Section 3:Building and Evaluating LLM Agents

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

from transformers import AutoTokenizer
class ChatAgent:
    def __init__(self, model):
        self.model = model
        self.tokenizer = AutoTokenizer.from_pretrained(model.model_id)

    def run(self, task):
        prompt = self.build_input(task)
        return self.model(prompt=prompt)

    def build_input(self, task):
        messages = [{"role": "user", "content": task}]
        return self.tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)

action_type="vanilla"
chat_agent_answers_files = {}
for task in ["simpleqa", "math", "gaia"]:
    chat_agent_answers_file = f"output/{model_id.replace('/', '__')}__{action_type}__{task}.jsonl"
    chat_agent = ChatAgent(model)
    answer_questions(task, chat_agent, action_type, chat_agent_answers_file, parallel_workers=8)
    chat_agent_answers_files[task] = chat_agent_answers_file
score_answers(chat_agent_answers_files.values())

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

from io import StringIO
import sys
import inspect
from typing import Any, Dict, Union, Optional, Tuple, List

class Tool:
    name: str
    description: str
    inputs: Dict[str, Dict[str, Union[str, type, bool]]]
    output_type: str

    def __init__(self, *args, **kwargs):
        self.is_initialized = False

    def forward(self, *args, **kwargs):
        return NotImplementedError("Write this method in your subclass of `Tool`.")

    def __call__(self, *args, **kwargs):
        outputs = self.forward(*args, **kwargs)
        return outputs

    def to_dict(self) -> dict:
        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": {
                    "type": "object",
                    "properties": self.inputs,
                    "required": list(self.inputs.keys()),
                },
            },
            "strict": True
        }


In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

from typing import Any

class FinalAnswerTool(Tool):
    name = "final_answer"
    description = "Provides a final answer to the given problem. Make sure to use this tool to provide the final answer at the end of the task."
    inputs = {"answer": {"type": "any", "description": "The final answer to the problem"}}
    output_type = "any"

    def forward(self, answer: Any) -> Any:
        return answer

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

final_answer_tool = FinalAnswerTool()
final_answer_tool("Hello, world!")

tools = [
    final_answer_tool.to_dict()
]

messages = [
    {"role": "user", "content": "What year was the municipality of Ramiriqu\u00ed, Boyac\u00e1, Colombia, founded?  Answer with only the final number."}
]
tokenizer = AutoTokenizer.from_pretrained(model.model_id)
print("="*10 + "Input Prompt" + "="*10 + "\n")
input_prompt = tokenizer.apply_chat_template(messages, tools=tools, tokenize=False, add_generation_prompt=True)
print("="*10 + "Response" + "="*10 + "\n")
response = model(tokenizer.apply_chat_template(messages, tools=tools, tokenize=False, add_generation_prompt=True))
print("\n\n" + response)


In [ ]:
class GoogleSearchTool(Tool):
    name = "web_search"
    description = """Performs a google web search for your query then returns a string of the top search results."""
    inputs = {
        "query": {"type": "string", "description": "The search query to perform."},
    }
    output_type = "string"

    def __init__(self, provider: str = "serper"):
        super().__init__()
        # TODO

    def forward(self, query: str) -> str:
        import requests

        # Register Google Search Api through https://serper.dev/
        # Use the organic key from search results to build this tool.

In [ ]:
import os
os.environ["SERPER_API_KEY"] = "APIKEY" # remember to remove your key for submission
google_search_tool = GoogleSearchTool()
google_search_tool("Space Exploration Technologies Corp.")

In [ ]:
!uv pip install markdownify

In [ ]:
class VisitWebpageTool(Tool):
    name = "visit_webpage"
    description = (
        "Visits a webpage at the given url and reads its content as a markdown string. Use this to browse webpages."
    )
    inputs = {
        "url": {
            "type": "string",
            "description": "The url of the webpage to visit.",
        }
    }
    output_type = "string"

    def __init__(self, max_output_length: int = 10000):
        super().__init__()
        self.max_output_length = max_output_length

    def forward(self, url: str) -> str:
        try:
            import re

            import requests
            from markdownify import markdownify
            from requests.exceptions import RequestException
        except ImportError as e:
            raise ImportError(
                "You must install packages `markdownify` and `requests` to run this tool: for instance run `pip install markdownify requests`."
            ) from e
        try:
            # TODO

        except requests.exceptions.Timeout:
            return "The request timed out. Please try again later or check the URL."
        except RequestException as e:
            return f"Error fetching the webpage: {str(e)}"
        except Exception as e:
            return f"An unexpected error occurred: {str(e)}"

In [ ]:
visit_page_tool = VisitWebpageTool()
visit_page_tool("https://en.wikipedia.org/wiki/SpaceX")

In [ ]:

import re
import json
from typing import List, Dict, Any, Callable, Optional

system_prompt = """
You are an expert assistant who can solve any task using tool calls. You will be given a task to solve as best you can.
To do so, you have been given access to some tools.

Here are the rules you should always follow to solve your task:
1. ALWAYS provide a tool call, else you will fail.
2. Always use the right arguments for the tools. Never use variable names as the action arguments, use the value instead.
3. Call a tool only when needed: do not call the search agent if you do not need information, try to solve the task yourself.
If no tool call is needed, use final_answer tool to return your answer.
4. Never re-do a tool call that you previously did with the exact same parameters.
"""


class ToolCallAgent:
    """
    A minimal agent that uses a language model to interact with tools
    based on the ToolCallingAgent pattern.
    """
    def __init__(
        self,
        model: Callable,
        tools: List[Tool],
        max_steps: int = 5,
        final_answer_tool_name: str = "final_answer",
        print_log: bool = True
    ):
        self.model = model
        self.tools_definitions = [tool.to_dict() for tool in tools]
        self.tool_executors = {tool.name: tool for tool in tools}
        self.max_steps = max_steps
        self.final_answer_tool_name = final_answer_tool_name

        self.log_func = lambda x: print(x) if print_log else lambda _: None

        # Validate that the final answer tool has an executor
        if self.final_answer_tool_name not in self.tool_executors:
            raise ValueError(f"Executor for final answer tool '{self.final_answer_tool_name}' not found in tool_executors.")

    def _execute_tool(self, tool_name: str, parsed_args: Any) -> str:
        executor = self.tool_executors[tool_name]
        try:
            self.log_func(f"Executing tool '{tool_name}'...")

            # TODO

            self.log_func(f"Tool '{tool_name}' executed.")
            # The final_answer tool itself returns the value, we don't stringify it here yet
            return tool_result
        except Exception as e:
            print(f"Error executing tool '{tool_name}' with args {parsed_args}: {e}")
            # Return the error message as the observation
            return f"Error during execution of tool '{tool_name}': {e}"

    def _parse_tool_arguments(self, content: str):
        """Try parse the tool calls."""
        # TODO

    def run(self, task: str) -> Any:
        """
        Runs the agent loop starting with the initial messages.

        Args:
            task: The initial user prompt.

        Returns:
            The final answer extracted from the final_answer tool call,
            or None if max_steps is reached or an error occurs.
        """


        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": task}
        ]

        self.log_func(f"Starting agent run with max_steps={self.max_steps}")
        self.log_func("Initial messages:")
        self.log_func(json.dumps(messages, indent=2))
        self.log_func("-" * 30)

        for step in range(self.max_steps):
            self.log_func(f"--- Step {step + 1} ---")
            self.log_func(json.dumps(messages, indent=2))

            # TODO
            # For guidance on handling and executing function calls, please refer to [this documentation](https://docs.sglang.ai/backend/function_calling.html).

            self.log_func("-" * 30) # Separator for next step

        self.log_func("Maximum steps reached without a final answer.")
        return None

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

def run_simpleqa_tool_calling_agent(task="simpleqa", model_id=model_id, action_type="tool_calling"):
    search_agent_answers_file = f"output/{model_id.replace('/', '__')}__{action_type}__{task}.jsonl"
    search_agent = ToolCallAgent(
        model=model,
        tools=[GoogleSearchTool(), VisitWebpageTool(), FinalAnswerTool()],
        max_steps=10
    )
    answer_questions(task, search_agent, action_type, search_agent_answers_file)
    df = score_answers([search_agent_answers_file])
    print(df)

run_simpleqa_tool_calling_agent()

In [ ]:
!wget -q -O local_python_executor.py https://raw.githubusercontent.com/ranpox/comp3361-spring2025/refs/heads/main/assignments/A3/local_python_executor.py

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

from local_python_executor import (
    BASE_BUILTIN_MODULES,
    BASE_PYTHON_TOOLS,
    evaluate_python_code,
)


class PythonInterpreterTool(Tool):
    name = "python_interpreter"
    description = "This is a tool for evaluating python code. It can be used to perform calculations. To generate valid Python code, first draft your code in Markdown using ```python```, then write it as a one-line string in the `code` field"
    inputs = {
        "code": {
            "type": "string",
            "description": "The python code to run in interpreter",
        }
    }
    output_type = "string"

    def __init__(self, *args, authorized_imports=None, **kwargs):
        if authorized_imports is None:
            self.authorized_imports = list(set(BASE_BUILTIN_MODULES))
        else:
            self.authorized_imports = list(set(BASE_BUILTIN_MODULES) | set(authorized_imports))
        self.inputs = {
            "code": {
                "type": "string",
                "description": (
                    "The code snippet to evaluate. All variables used in this snippet must be defined in this same snippet, "
                    f"else you will get an error. This code can only import the following python libraries: {self.authorized_imports}."
                ),
            }
        }
        self.base_python_tools = BASE_PYTHON_TOOLS
        self.python_evaluator = evaluate_python_code
        super().__init__(*args, **kwargs)

    def forward(self, code: str) -> str:
        state = {}
        output = str(
            self.python_evaluator(
                code,
                state=state,
                static_tools=self.base_python_tools,
                authorized_imports=self.authorized_imports,
            )  # The second element is boolean is_final_answer
        )
        return f"Stdout:\n{str(state['_print_outputs'])}\nOutput: {output}"

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

from eval_utils import score_answers

def run_math_tool_calling_agent(task="math", model_id=model_id, action_type="tool_calling"):
    math_agent_answers_file = f"output/{model_id.replace('/', '__')}__{action_type}__{task}.jsonl"
    math_agent = ToolCallAgent(
        model=model,
        tools=[PythonInterpreterTool(authorized_imports=["numpy", "sympy"]), FinalAnswerTool()],
        max_steps=10
    )
    answer_questions(task, math_agent, action_type, math_agent_answers_file)
    df = score_answers([math_agent_answers_file])
    print(df)

run_math_tool_calling_agent()

In [ ]:
######################################################
#  The following helper code is given to you.
######################################################

from eval_utils import score_answers

def run_gaia_tool_calling_agent(task="gaia", model_id=model_id, action_type="tool_calling"):
    research_agent_answers_file = f"output/{model_id.replace('/', '__')}__{action_type}__{task}.jsonl"
    research_agent = ToolCallAgent(
        model=model,
        tools=[GoogleSearchTool(), VisitWebpageTool(), PythonInterpreterTool(authorized_imports=["numpy", "sympy"]), FinalAnswerTool()],
        max_steps=10
    )
    answer_questions(task, research_agent, action_type, research_agent_answers_file)
    df = score_answers([research_agent_answers_file])
    print(df)

run_gaia_tool_calling_agent()

| Task        | Action Type | Performance      |
|-------------|-----|-----------------|
| SimpleQA       | vanilla  |  |
| SimpleQA       | tool calling  |  |
| MATH       | vanilla  |  |
| MATH       | tool calling  |  |
| GAIA       | vanilla  |  |
| GAIA       | tool calling  |  |